source/fab/application.py

```diff
@@ -58,6 +87,89 @@ def run(self, source: Path):
                 print('    found in: ' + str(filename))
                 print('    depends on: ' + str(fortran_db.depends_on(unit)))
 
+        # Start with the top level program unit
+        unit_to_process = [self._target]
+
+        # Initialise linker
+        if self._exec_name != "":
+            executable = Path(self._workspace) / self._exec_name
+        else:
+            executable = \
+                Path(self._workspace / self._target).with_suffix(".exe")
+
+        flags = self._command_flags_map.get(FortranLinker, [])
+        link_command = FortranLinker(self._workspace, flags, executable)
+
+        processed_units = []
+
+        while unit_to_process:
+            # Pop pending items from start of the list
+            unit = unit_to_process.pop(0)
+            if unit in processed_units:
+                continue
+
+            dependencies = fortran_db.depends_on(unit)
+            unit_to_process.extend(dependencies)
+
+            filenames = fortran_db.filenames_from_program_unit(unit)
+            # TODO: For now we simply can't handle this returning
+            #       multiple names; later we may be able to deal with
+            #       this via extra information in the configuration
+            if len(filenames) == 1:
+                filename = filenames[0]
+            else:
+                raise ValueError("Duplicate program unit found")
+
+            # Construct names of any expected module files to
+            # pass to the compiler constructor
+            mod_files = [Path(self._workspace /
+                              dependee).with_suffix('.mod')
+                         for dependee in dependencies]
+
+            compiler_class = self._compiler_map[filename.suffix]
+
+            if issubclass(compiler_class, FortranCompiler):
```

Comment:
I expect you're right - I kind of thought it would eventually be a `Compiler` subclass of `Task` into which all compilers would fall - but then again at some level we'll eventually be dealing with both C and Fortran, at which point I guess there might need to be some branching here anyway?